projects/igniteui-angular/src/lib/directives/mask/mask.directive.ts

```diff
@@ -105,264 +87,214 @@ export class IgxMaskDirective implements OnInit, ControlValueAccessor {
             this._stopPropagation = false;
             return;
         }
 
-        if (this._paste) {
-            this._paste = false;
-
-            const clipboardData = this.value.substring(this._cursorOnPaste, this.getCursorPosition());
-            this.value = this.maskParser.parseValueByMaskUponCopyPaste(
-                this._valOnPaste, this._maskOptions, this._cursorOnPaste, clipboardData, this._selection);
-
-            this.setCursorPosition(this.maskParser.cursor);
-        } else {
-            const currentCursorPos = this.getCursorPosition();
+        const valueToParse = event.data || this.valueToParse;
+        const hasDeleteAction = (this._key === KEYCODES.BACKSPACE) || (this._key === KEYCODES.DELETE);
+        this._cursor = this._hasDropAction ? this.selectionStart : this.updateCursorOnPasteOrDelete(hasDeleteAction, event);
+        const parsedData = this.maskParser.parseMaskValue(
+            this._oldVal, this.inputValue, this.maskOptions, this._cursor,
+            valueToParse, this._selection, hasDeleteAction);
+        this.inputValue = parsedData.value;
+        this.setSelectionRange(parsedData.cursor);
+
+        const rawVal = this.maskParser.restoreValueFromMask(this.inputValue, this.maskOptions);
+        this._dataValue = this.includeLiterals ? this.inputValue : rawVal;
+        this._onChangeCallback(this._dataValue);
+
+        this.onValueChange.emit({ rawValue: rawVal, formattedValue: this.inputValue });
+        this.afterInput();
+    }
 
-            this.maskParser.data = (this._key === KEYS.BACKSPACE) || (this._key === KEYS.DELETE);
+    /** @hidden */
+    @HostListener('paste')
+    public onPaste(): void {
+        this._oldVal = this.inputValue;
+        this._cursor = this.selectionStart;
+    }
 
-            this.value = this._selection && this._selection !== 0 ?
-                this.maskParser.parseValueByMaskUponSelection(this.value, this._maskOptions, currentCursorPos - 1, this._selection) :
-                this.maskParser.parseValueByMask(this.value, this._maskOptions, currentCursorPos - 1);
+    /** @hidden */
+    @HostListener('focus', ['$event'])
+    public onFocus(event: FocusEvent): void {
+        this.showMask((event.target as HTMLInputElement).value);
+    }
 
-            this.setCursorPosition(this.maskParser.cursor);
+    /** @hidden */
+    @HostListener('blur', ['$event.target.value'])
+    public onBlur(value: string): void {
+        if (this.displayValuePipe) {
+            this.inputValue = this.displayValuePipe.transform(value);
+        } else if (value === this.maskParser.parseMask(null, this.maskOptions)) {
+            this.inputValue = '';
         }
+        this._onTouchedCallback();
+    }
 
-        const rawVal = this.maskParser.restoreValueFromMask(this.value, this._maskOptions);
-
-        this.dataValue = this.includeLiterals ? this.value : rawVal;
-        this._onChangeCallback(this.dataValue);
+    /** @hidden */
+    @HostListener('dragover')
```

Comment:
`dragover` is a very chatty event, `dragenter` should do just fine. Should also add `dragleave` to reset the state